### PR TITLE
Non-record: Value Residual (-0.015 BPB) + Gated Attention (-0.003 BPB) with ablations

### PR DESCRIPTION
**Pull request by anantdgoel**
## Non-record: Value Residual (-0.015 BPB) + Gated Attention (-0.003 BPB) with ablations

**val_bpb: 1.4525** (sliding window, stride=128, GA+VR combined) | **13.2 MB** | 1xRTX3090, 1000 steps, 131K batch

Two novel architecture modifications and one negative result. Sharing validated techniques with controlled ablation data.

### Contributions

1. **Value Residual (ResFormer)** -- -0.015 BPB. Cache V vectors from layer 0, mix into all subsequent layers via learnable scalars. 18 params total. arXiv:2410.17897 (ACL 2025). Enable: `VALUE_RESIDUAL=1`.

2. **Gated Attention** -- -0.003 BPB. Per-head sigmoid gate after SDPA output, eliminating attention sinks. ~37K params. arXiv:2505.06708 (NeurIPS 2025 Best Paper). Enable: `GATED_ATTENTION=1`.

3. **PPM-C Context Mixer** -- +0.0018 BPB (negative result). Classical compression blended with neural softmax. Dilutes predictions on SmearGate+BigramHash models.

The two positive techniques **stack additively** for **-0.017 BPB** combined.

### Ablation Results

v1024 9L 2xMLP, SmearGate + BigramHash + OrthoInit + WD 0.04, 131K batch, 1000 steps.

| Config | Sliding BPB | Delta vs Control |
|--------|-------------|------------------|
| Control | 1.4697 | -- |
| Gated Attention only | 1.4665 | -0.0032 |
| **Value Residual only** | **1.4546** | **-0.0151** |
| **GA + VR combined** | **1.4525** | **-0.0172** |
| PPM-C (eval-only) | 1.2900 | +0.0018 (worse) |

A production run (11L MLP3x + full community stack + VR + GA, 9500 steps) is in progress. Results in a follow-up submission if competitive.

### Files

- `README.md` -- Full writeup with technique details and reproducibility
- `submission.json` -- Metadata
- `train_gpt.py` -- Training script with Value Residual, Gated Attention, XSA, EMA, Partial RoPE, LN Scale

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #413

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Non-record: Value Residual (-0.015 BPB) + Gated Attention (-0.003 BPB) with ablations  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | True | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 3 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~1659 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** False  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- TTT (Test-Time Training) integration detected
- Custom tokenizer patterns found (3 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #413

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Non-record: Value Residual (-0.015 BPB) + Gated Attention (-0.003 BPB) with ablations  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | True | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 3 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~1659 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** False  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- TTT (Test-Time Training) integration detected
- Custom tokenizer patterns found (3 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Value Residual (-0.015 BPB) + Gated Attention (-0.003 BPB) with ablations

**Compliance flag: Pre-Quant TTT violation (multi-epoch SGD on val_tokens before scoring)**


## PR 413 — Non-record: Value Residual + Gated Attention (anantdgoel)

**Head SHA:** c0a2939233e4fd8dc56c924f33a41e390d0fee2e  
**Track:** non-record-16mb  
**Reported val_bpb:** 1.4525 (pre_quant_val_bpb: 1.4525)  
**Hardware:** 1×RTX3090 (ablations), 1×A6000 (validation)

---

### Check 1: N-gram Family Bug (target token in hash key)

**CLEAN.** The `BigramHash` at line 930–936 computes:

```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```

Position `i` hashes tokens `[i]` (current) and `[i-1]` (previous). The current token `t[i]` is the token being *embedded*, not the *target* (next token). This is causally clean — no look-ahead. This is the legal `BigramHash` pattern (same as baseline). No N-gram family bug.

---

### Check 2: Pre-Quant TTT (multi-epoch AdamW on val_tokens without score-first)

**CLOSE TRIGGER.** The SGD TTT at lines 429–442 (`eval_val_sgd_ttt`) runs a two-phase procedure:

- **Phase 1 (Adapt):** multi-epoch SGD over all `val_tokens` — iterates `args.sgd_ttt_epochs` (default: 2) passes over the entire validation set, computing loss and updating weights.
- **Phase 2 (Score):** sliding-window scoring on the *already-adapted* model.

This is adapt-first, score-second on the full val set — the banned Pre-Quant TTT pattern. The optimizer is SGD (not AdamW), but the structure is identical: gradient updates on val_tokens before scoring, across multiple epochs, with no score-first-per-chunk interleaving. The SGD vs. AdamW distinction does not change the violation — the invariant is about adapting on val data before scoring, which this does unambiguously.

**Verdict:** CLOSE — Pre-Quant TTT violation (multi-epoch SGD on val_tokens before scoring).

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** Recommend **CLOSE** unless the author restructures the flagged code.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
